File: src/tools/shared/dbHealthcheck.ts
import type { ServiceSpec } from "../../types/stack";

export interface DefaultHealthcheck {
  imagePattern: RegExp;
  healthcheck: {
    test: string | string[];
    interval: string;
    timeout: string;
    retries: number;
    start_period: string;
  };
}

export const DEFAULT_DB_HEALTHCHECKS: DefaultHealthcheck[] = [
  {
    imagePattern: /^postgres(:|$)/,
    healthcheck: {
      test: ["CMD-SHELL", "pg_isready -U \${POSTGRES_USER:-postgres}"],
      interval: "10s",
      timeout: "5s",
      retries: 5,
      start_period: "30s",
    },
  },
  {
    imagePattern: /^mysql(:|$)/,
    healthcheck: {
      test: ["CMD", "mysqladmin", "ping", "-h", "localhost"],
      interval: "10s",
      timeout: "5s",
      retries: 5,
      start_period: "30s",
    },
  },
  {
    imagePattern: /^mariadb(:|$)/,
    healthcheck: {
      test: ["CMD", "healthcheck.sh", "--connect", "--innodb_initialized"],
      interval: "10s",
      timeout: "5s",
      retries: 5,
      start_period: "30s",
    },
  },
  {
    imagePattern: /^mongo(:|$)/,
    healthcheck: {
      test: ["CMD", "mongosh", "--eval", "db.adminCommand('ping')"],
      interval: "10s",
      timeout: "5s",
      retries: 5,
      start_period: "30s",
    },
  },
  {
    imagePattern: /^redis(:|$)/,
    healthcheck: {
      test: ["CMD", "redis-cli", "ping"],
      interval: "10s",
      timeout: "3s",
      retries: 5,
      start_period: "5s",
    },
  },
];

/**
 * Auto-injects default healthchecks for databases in the planned stack services
 * and updates dependent services' depends_on conditions to service_healthy.
 */
export function injectDbHealthchecks(services: Record<string, ServiceSpec>): {
  injectedCount: number;
  updatedDepsCount: number;
} {
  let injectedCount = 0;
  let updatedDepsCount = 0;

  const dbServices = new Set<string>();

  // 1. Inject healthchecks for DBs (both catalog and custom images)
  for (const [name, spec] of Object.entries(services)) {
    const rule = DEFAULT_DB_HEALTHCHECKS.find((r) => r.imagePattern.test(spec.image));
    if (rule) {
      dbServices.add(name);
      if (!spec.healthcheck) {
        spec.healthcheck = { ...rule.healthcheck };
        injectedCount++;
      }
    }
  }

  // 2. Update depends_on for services that depend on the injected DB services
  for (const [name, spec] of Object.entries(services)) {
    if (!spec.depends_on) continue;

    if (Array.isArray(spec.depends_on)) {
      // Convert array depends_on to record with service_healthy condition for DB services
      const newDependsOn: Record<
        string,
        { condition: "service_started" | "service_healthy" | "service_completed_successfully" }
      > = {};
      let changed = false;

      for (const dep of spec.depends_on) {
        if (dbServices.has(dep)) {
          newDependsOn[dep] = { condition: "service_healthy" };
          changed = true;
        } else {
          newDependsOn[dep] = { condition: "service_started" };
        }
      }

      if (changed) {
        spec.depends_on = newDependsOn;
        updatedDepsCount++;
      }
    } else if (typeof spec.depends_on === "object") {
      // Check if any DB service is in the record and doesn't already demand service_healthy
      let changed = false;
      const newDependsOn = { ...spec.depends_on };

      for (const dep of Object.keys(newDependsOn)) {
        if (dbServices.has(dep)) {
          const currentDep = newDependsOn[dep];
          // Do NOT override service_completed_successfully or other explicit conditions
          if (
            currentDep &&
            currentDep.condition !== "service_healthy" &&
            currentDep.condition !== "service_completed_successfully"
          ) {
            newDependsOn[dep] = { ...currentDep, condition: "service_healthy" };
            changed = true;
          }
        }
      }

      if (changed) {
        spec.depends_on = newDependsOn;
        updatedDepsCount++;
      }
    }
  }

  return { injectedCount, updatedDepsCount };
}

File: servers/docker-mcp-server/src/docker_mcp_server/tools/shared/db_healthcheck.py
"""Default database healthchecks and depends_on upgrades.

Parity: ``src/tools/shared/dbHealthcheck.ts``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

from docker_mcp_server.types.stack import HealthcheckSpec, ServiceSpec

DependsOnConditionName = Literal[
    "service_started", "service_healthy", "service_completed_successfully"
]


@dataclass(frozen=True)
class DefaultHealthcheck:
    image_pattern: re.Pattern[str]
    healthcheck: HealthcheckSpec


DEFAULT_DB_HEALTHCHECKS: list[DefaultHealthcheck] = [
    DefaultHealthcheck(
        image_pattern=re.compile(r"^postgres(:|$)"),
        healthcheck=HealthcheckSpec.model_validate(
            {
                "test": ["CMD-SHELL", "pg_isready -U ${POSTGRES_USER:-postgres}"],
                "interval": "10s",
                "timeout": "5s",
                "retries": 5,
                "start_period": "30s",
            }
        ),
    ),
    DefaultHealthcheck(
        image_pattern=re.compile(r"^mysql(:|$)"),
        healthcheck=HealthcheckSpec.model_validate(
            {
                "test": ["CMD", "mysqladmin", "ping", "-h", "localhost"],
                "interval": "10s",
                "timeout": "5s",
                "retries": 5,
                "start_period": "30s",
            }
        ),
    ),
    DefaultHealthcheck(
        image_pattern=re.compile(r"^mariadb(:|$)"),
        healthcheck=HealthcheckSpec.model_validate(
            {
                "test": ["CMD", "healthcheck.sh", "--connect", "--innodb_initialized"],
                "interval": "10s",
                "timeout": "5s",
                "retries": 5,
                "start_period": "30s",
            }
        ),
    ),
    DefaultHealthcheck(
        image_pattern=re.compile(r"^mongo(:|$)"),
        healthcheck=HealthcheckSpec.model_validate(
            {
                "test": ["CMD", "mongosh", "--eval", "db.adminCommand('ping')"],
                "interval": "10s",
                "timeout": "5s",
                "retries": 5,
                "start_period": "30s",
            }
        ),
    ),
    DefaultHealthcheck(
        image_pattern=re.compile(r"^redis(:|$)"),
        healthcheck=HealthcheckSpec.model_validate(
            {
                "test": ["CMD", "redis-cli", "ping"],
                "interval": "10s",
                "timeout": "3s",
                "retries": 5,
                "start_period": "5s",
            }
        ),
    ),
]


def inject_db_healthchecks(
    services: dict[str, ServiceSpec],
) -> dict[str, int]:
    """Inject healthchecks for DB images and upgrade depends_on conditions."""
    injected_count = 0
    updated_deps_count = 0
    db_services: set[str] = set()

    for name, spec in services.items():
        rule = next(
            (r for r in DEFAULT_DB_HEALTHCHECKS if r.image_pattern.search(spec.image)),
            None,
        )
        if rule is not None:
            db_services.add(name)
            if spec.healthcheck is None:
                spec.healthcheck = rule.healthcheck.model_copy(deep=True)
                injected_count += 1

    for _name, spec in services.items():
        if spec.depends_on is None:
            continue

        if isinstance(spec.depends_on, list):
            new_depends_on: dict[str, dict[str, DependsOnConditionName]] = {}
            changed = False
            for dep in spec.depends_on:
                if dep in db_services:
                    new_depends_on[dep] = {"condition": "service_healthy"}
                    changed = True
                else:
                    new_depends_on[dep] = {"condition": "service_started"}
            if changed:
                spec.depends_on = new_depends_on
                updated_deps_count += 1
        elif isinstance(spec.depends_on, dict):
            changed = False
            new_depends_on = dict(spec.depends_on)
            for dep in list(new_depends_on.keys()):
                if dep not in db_services:
                    continue
                current_dep = new_depends_on[dep]
                if (
                    isinstance(current_dep, dict)
                    and current_dep.get("condition") != "service_healthy"
                    and current_dep.get("condition") != "service_completed_successfully"
                ):
                    new_depends_on[dep] = {**current_dep, "condition": "service_healthy"}
                    changed = True
            if changed:
                spec.depends_on = new_depends_on
                updated_deps_count += 1

    return {"injected_count": injected_count, "updated_deps_count": updated_deps_count}


__all__ = [
    "DEFAULT_DB_HEALTHCHECKS",
    "DefaultHealthcheck",
    "inject_db_healthchecks",
]
